rhel/vex/fetcher.go

```diff
@@ -0,0 +1,288 @@
+package vex
+
+import (
+	"archive/tar"
+	"bytes"
+	"context"
+	"encoding/csv"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"path"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/klauspost/compress/snappy"
+	"github.com/klauspost/compress/zstd"
+	"github.com/quay/zlog"
+
+	"github.com/quay/claircore/libvuln/driver"
+	"github.com/quay/claircore/pkg/tmp"
+)
+
+// A Changes is a series of Entries indicating where to find resources in the
+// changes.csv file.
+type Changes []Entry
+
+// Entry is an entry in a changes.csv file.
+type Entry struct {
+	// This path should be parsed in the context of the manifest's URL.
+	Path        string
+	updatedTime time.Time
+}
+
+func (u *VEXUpdater) Fetch(ctx context.Context, hint driver.Fingerprint) (io.ReadCloser, driver.Fingerprint, error) {
+	ctx = zlog.ContextWithValues(ctx, "component", "rhel/VEXUpdater.Fetch")
+	fp, err := ParseFingerprint(string(hint))
+	if err != nil {
+		return nil, hint, err
+	}
+
+	f, err := tmp.NewFile("", "rhel-vex.")
+	if err != nil {
+		return nil, hint, err
+	}
+
+	cw := snappy.NewBufferedWriter(f)
+
+	var success bool
+	defer func() {
+		if err := cw.Close(); err != nil {
+			zlog.Warn(ctx).Err(err).Msg("unable to close snappy writer")
+		}
+		if success {
+			if _, err := f.Seek(0, io.SeekStart); err != nil {
+				zlog.Warn(ctx).
+					Err(err).
+					Msg("unable to seek file back to start")
+			}
+		} else {
+			if err := f.Close(); err != nil {
+				zlog.Warn(ctx).Err(err).Msg("unable to close spool")
+			}
+		}
+	}()
+
+	if fp.changesEtag == "" { // Used to inform whether this is the first run.
+		// We need to go after the full corpus of vulnerabilities
+		// First we target the archive_latest.txt file
+		latestURI, err := u.url.Parse(latestFile)
+		if err != nil {
+			return nil, hint, err
+		}
+		latestReq, err := http.NewRequestWithContext(ctx, http.MethodGet, latestURI.String(), nil)
+		if err != nil {
+			return nil, hint, err
+		}
+		latestRes, err := u.client.Do(latestReq)
+		if latestRes != nil {
+			defer latestRes.Body.Close()
+		}
+		if err != nil {
+			return nil, hint, err
+		}
+		if latestRes.StatusCode != http.StatusOK {
+			return nil, hint, fmt.Errorf("unexpected response from archive_latest.txt: %s", latestRes.Status)
+		}
+
+		body, err := io.ReadAll(latestRes.Body) // Fine to use as expecting small number of bytes.
+		if err != nil {
+			return nil, hint, err
+		}
+
+		compressedFilename := string(body)
+		zlog.Debug(ctx).
+			Str("filename", compressedFilename).
+			Msg("requesting latest compressed file")
+
+		// Decided against parsing the filename to allow the implementation
```

Comment:
I think I added this as the compressed filename is when it was generated but I think it's kind of moot since prodsec added the `archive_latest.txt` endpoint, will remove comment.